crates/fuel-core/src/service/adapters/fuel_gas_price_provider.rs

```diff
@@ -0,0 +1,228 @@
+        let mut total_reward = self.total_reward.borrow_mut();
+        let mut total_cost = self.total_cost.borrow_mut();
+        *totaled_block_height = block_height;
+        *total_reward += reward;
+        *total_cost += cost;
+    }
+
+    fn totaled_block_height(&self) -> BlockHeight {
+        *self.totaled_block_height.borrow()
+    }
+}
+
+impl<FB, DA, A, GP> FuelGasPriceProvider<FB, DA, A, GP>
+where
+    FB: FuelBlockHistory,
+    DA: DARecordingCostHistory,
+    A: GasPriceAlgorithm,
+    GP: GasPriceHistory,
+{
+    fn inner_gas_price(&self, requested_block_height: BlockHeight) -> Result<u64> {
+        let latest_block = self
+            .block_history
+            .latest_height()
+            .map_err(Error::UnableToGetLatestBlockHeight)?;
+        if latest_block > requested_block_height {
+            let gas_prices = self
+                .gas_price_history
+                .gas_prices(requested_block_height)
+                .map_err(Error::UnableToGetGasPrices)?
+                .ok_or(Error::GasPricesNotFoundForBlockHeight(
+                    requested_block_height,
+                ))?;
+            Ok(gas_prices.total())
+        } else if Self::asking_for_next_block(latest_block, requested_block_height) {
+            let new_gas_prices = self.calculate_new_gas_price(latest_block)?;
+            self.gas_price_history
+                .store_gas_prices(latest_block, &new_gas_prices)
+                .map_err(Error::UnableToStoreGasPrices)?;
+            Ok(new_gas_prices.total())
+        } else {
+            Err(Error::RequestedBlockHeightTooHigh {
+                requested: requested_block_height,
+                latest: latest_block,
+            })
+        }
+    }
+
+    fn asking_for_next_block(
+        latest_block: BlockHeight,
+        block_height: BlockHeight,
+    ) -> bool {
+        *latest_block + 1 == *block_height
+    }
+
+    fn calculate_new_gas_price(
+        &self,
+        latest_block_height: BlockHeight,
+    ) -> Result<GasPrices> {
+        let previous_gas_prices = self
+            .gas_price_history
+            .gas_prices(latest_block_height)
+            .map_err(Error::UnableToGetGasPrices)?
+            .ok_or(Error::GasPricesNotFoundForBlockHeight(latest_block_height))?;
+        let latest_da_recorded_block = self
+            .da_recording_cost_history
+            .latest_height()
+            .map_err(Error::UnableToGetLatestBlockHeight)?;
+
+        let new_da_gas_price = if latest_da_recorded_block
+            <= self.profitablility_totals.totaled_block_height()
+        {
+            previous_gas_prices.da_gas_price
+        } else {
+            self.update_totals(latest_block_height)?;
+            self.algorithm.calculate_da_gas_price(
+                previous_gas_prices.da_gas_price,
+                self.total_reward(),
+                self.total_cost(),
+            )
+        };
+
+        let block_fullness = self
```

Comment:
I think the previous block's fullness should be part of the `GasProviderService`. On each block imported from the block importer, the service will update its internal fullness variables and dump new values into the database.

Reply:
I'm not sure what you mean `GasProviderService`. Right now the producer, the sync, and graphql each get a copy of the `impl GasPriceProvider for StaticGasPrice`.

That's how I imagine it will still work, but just with this new `FuelGasPriceProvider` instead of `StaticGasPrice`. I'm assuming that all the `GasPriceProvider` needs is a `BlockHeight` and then it has access to all the DBs it needs to assemble the gas price. I don't know why it needs to store `block_fullness` when it can just look up the block's data in its DB.